### PR TITLE
feat: overflow theorems

### PR DESCRIPTION
**Pull request by AnotherAlexHere**
Added new theorems for overflow for addition, subtraction, multiplication and division.

### CODE REVIEW

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 1]
```diff
@@ -302,7 +302,113 @@ theorem slt_iff_adc :
 
 theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
     (x ≤ₛ 0) ↔ (BitVec.carry w (2 ^ (w - 1)) (-(x + 2 ^ (w - 1)))) false := by
-  sorry
+<<<<<<< Updated upstream
+  sorry
+=======
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]: This is called a conflict marker. Git uses this to show when you've had a merge conflict. Just delete this.

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 2]
```diff
@@ -3,9 +3,12 @@ import Init.Data.BitVec.Lemmas
 import Init.Data.BitVec.Bitblast
 import SSA.Projects.InstCombine.ForMathlib
 import SSA.Projects.InstCombine.ForStd
+import SSA.Projects.InstCombine.TacticAuto
 import Init.Data.Nat.Bitwise.Basic
 import Init.Data.Nat.Bitwise.Lemmas
 
+set_option linter.unusedTactic false
+set_option linter.unreachableTactic false
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]: These changes are unrelated to this PR. You should remove them and add them to a different PR.

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 3]
```diff
@@ -3,9 +3,12 @@ import Init.Data.BitVec.Lemmas
 import Init.Data.BitVec.Bitblast
 import SSA.Projects.InstCombine.ForMathlib
 import SSA.Projects.InstCombine.ForStd
+import SSA.Projects.InstCombine.TacticAuto
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]: This change is unrelated to this PR.

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 4]
```diff
@@ -304,6 +307,89 @@ theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
...
+theorem add_not_iff_ult :
+    AdditionNoOverflows? x (~~~ y) ↔ x ≤ᵤ y := by
+  sorry
+
+theorem add_not_iff_ule_sub_sub :
+    AdditionNoOverflows? x (~~~ y) ↔ x ≤ᵤ x - y - 1 := by
+  sorry
+
+def UnsignedMultiplicationOverflows? (x y : BitVec w) : Prop := x.toNat * y.toNat > 2 ^ w
+def SignedMultiplicationOverflows? (x y : BitVec w) : Prop := x.toNat * y.toNat > 2 ^ (w - 1)
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]: I think this should be x.toInt, but I'm not sure. Check with the Hacker's Delight book.

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 5]
```diff
@@ -304,6 +307,89 @@ theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
...
+
+theorem mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
+    (x * y) / z ≠ x → UnsignedMultiplicationOverflows? x y := by
+  sorry
+
+theorem lt_and_eq_neg_pow_two_or_mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
+    (y < 0) && (x.toInt = - 2 ^ 31) || ((x * y) / z ≠ x) → SignedMultiplicationOverflows? x y := by
+  sorry
+
+def NumberOfLeadingZeros {w : Nat} (x : BitVec w) : Nat :=
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]: This theorem name should start with a lowercase letter

https://leanprover-community.github.io/contribute/naming.html

**Review by AtticusKuhn — APPROVED** [review]
Looks good to me.

File: SSA/Projects/InstCombine/HackersDelight.lean (hunk 5 above)
Comment (AtticusKuhn) [reply SSA/Projects/InstCombine/HackersDelight.lean]: Via rule 3 and then rule 4.

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 6]
```diff
@@ -304,6 +307,89 @@ theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
...
+theorem signed_mul_overflow_iff_mul_neq_mul_RightShift {x y : BitVec 64} :
+    SignedMultiplicationOverflows? x y ↔ First32Bits (x * y) ≠ Last32Bits (x * y) >>> 31 := by
+  sorry
+
+theorem mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
+    (x * y) / z ≠ x → UnsignedMultiplicationOverflows? x y := by
+  sorry
+
+theorem lt_and_eq_neg_pow_two_or_mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
+    (y < 0) && (x.toInt = - 2 ^ 31) || ((x * y) / z ≠ x) → SignedMultiplicationOverflows? x y := by
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]:
```suggestion
    (y < 0) ∧ (x.toInt = - 2 ^ 31) ∨  ((x * y) / z ≠ x) → SignedMultiplicationOverflows? x y := by
```

It's better to use logic/prop notation than Bool notation.

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 7]
```diff
@@ -304,6 +307,89 @@ theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
...
+
+theorem signed_mul_overflow_iff_mul_neq_mul_RightShift {x y : BitVec 64} :
+    SignedMultiplicationOverflows? x y ↔ First32Bits (x * y) ≠ Last32Bits (x * y) >>> 31 := by
+  sorry
+
+theorem mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
+    (x * y) / z ≠ x → UnsignedMultiplicationOverflows? x y := by
+  sorry
+
+theorem lt_and_eq_neg_pow_two_or_mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean:357]:
```suggestion
theorem lt_and_eq_neg_pow_two_or_mul_div_neq_imp_unsigned_mul_overflow (h : y ≠ 0) :
```

Surely "y.toNat ≠ 0" is logically equivalent to "y ≠ 0".

File: SSA/Projects/InstCombine/HackersDelight.lean (hunk 6 above)
Comment (AtticusKuhn) [reply SSA/Projects/InstCombine/HackersDelight.lean]: To see for yourself why this is important,

try

```lean
#check (1 = 1) && (1 = 1)
#check (1 = 1) ∧ (1 = 1)
```

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 8]
```diff
@@ -304,6 +307,89 @@ theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
...
+    NumberOfLeadingZeros x + NumberOfLeadingZeros y ≤ 30 ↔ UnsignedMultiplicationOverflows? x y := by
+  sorry
+
+def DivisionOverflows? (x y : BitVec w) : Prop := sorry
+
+theorem div_overflow_iff_eq_zero_or_eq_neg_pow_two_and_eq_neg :
+    DivisionOverflows? x y ↔ y = 0 || ((x.toInt = -2 ^ 31) && y = -1) := by
+  sorry
+
+theorem div_overflow_iff_neq_and_ult_LeftShift {x y : BitVec 64} (h : y.toNat = Last32Bits y) :
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]: What type is this equality on? Nats or bitvecs?

"y.toNat = Last32Bits y"

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 9]
```diff
@@ -304,6 +307,89 @@ theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
...
+  sorry
+
+theorem nlz_add_nlz_le_unsigned_mul_overflow {x y : BitVec 64} :
+    NumberOfLeadingZeros x + NumberOfLeadingZeros y ≤ 30 ↔ UnsignedMultiplicationOverflows? x y := by
+  sorry
+
+def DivisionOverflows? (x y : BitVec w) : Prop := sorry
+
+theorem div_overflow_iff_eq_zero_or_eq_neg_pow_two_and_eq_neg :
+    DivisionOverflows? x y ↔ y = 0 || ((x.toInt = -2 ^ 31) && y = -1) := by
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]: Same comment here as above.

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 10]
```diff
@@ -304,6 +307,89 @@ theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
...
+
+theorem le_nlz_add_nlz_not_unsigned_mul_overflow {x y : BitVec 64} :
+    32 ≤ NumberOfLeadingZeros x + NumberOfLeadingZeros y ↔ ¬ UnsignedMultiplicationOverflows? x y := by
+  sorry
+
+theorem nlz_add_nlz_le_unsigned_mul_overflow {x y : BitVec 64} :
+    NumberOfLeadingZeros x + NumberOfLeadingZeros y ≤ 30 ↔ UnsignedMultiplicationOverflows? x y := by
+  sorry
+
+def DivisionOverflows? (x y : BitVec w) : Prop := sorry
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]: Is this signed division or unsigned division?

Comment (AtticusKuhn) [reply SSA/Projects/InstCombine/HackersDelight.lean]: If it's signed division, the answer you're looking for is 

```lean
y = 0 ∨  w ≠ 1 ∧ x = intMin w ∧  y = -1

```

Comment (AtticusKuhn) [reply SSA/Projects/InstCombine/HackersDelight.lean]: This came from SSA/Projects/InstCombine/LLVM/Semantics.lean

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 11]
```diff
@@ -143,13 +143,13 @@ theorem and_ule_not_xor :
     x &&& y ≤ᵤ ~~~(x ^^^ y) := by
   sorry
 
-def AdditionNoOverflows? (x y : BitVec w) : Prop := (x.adc y false).1
+def additionNoOverflows? (x y : BitVec w) : Prop := (x.adc y false).1
```
Comment (AtticusKuhn) [inline SSA/Projects/InstCombine/HackersDelight.lean]: This should be capitalised.

Comment (AtticusKuhn) [reply SSA/Projects/InstCombine/HackersDelight.lean]: https://leanprover-community.github.io/contribute/naming.html)](https://leanprover-community.github.io/contribute/naming.html

Rule 2

Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/HackersDelight.lean]: You can always remove it from the queue and fix things. 😄

File: SSA/Projects/InstCombine/HackersDelight.lean [hunk 12]
```diff
@@ -304,6 +304,89 @@ theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
...
+theorem add_not_iff_ult :
+    additionNoOverflows? x (~~~ y) ↔ x ≤ᵤ y := by
+  sorry
+
+theorem add_not_iff_ule_sub_sub :
+    additionNoOverflows? x (~~~ y) ↔ x ≤ᵤ x - y - 1 := by
+  sorry
+
+def unsignedMultiplicationOverflows? (x y : BitVec w) : Prop := x.toNat * y.toNat > 2 ^ w
+def signedMultiplicationOverflows? (x y : BitVec w) : Prop := x.toInt * y.toInt > 2 ^ (w - 1)
```
Comment (Equilibris) [inline SSA/Projects/InstCombine/HackersDelight.lean]: Please follow naming conventions here as well

**Review by Equilibris — APPROVED** [review]

**Pull request merged**